Sources/OpenCombine/Publishers/Publishers.Map.swift

```diff
@@ -32,70 +43,121 @@ extension Publishers {
 
-    init(downstream: Downstream, transform: @escaping Transform) {
-        _downstream = downstream
-        _transform = transform
+    init(downstream: Downstream, genericTransform: @escaping GenericTransform) {
+        _genericTransform = genericTransform
+        super.init(downstream: downstream)
     }
 
+    var description: String { return "Map" }
+
     func receive(subscription: Subscription) {
-        _upstreamSubscription = subscription
-        _downstream.receive(subscription: self)
+        upstreamSubscription = subscription
+        downstream.receive(subscription: self)
     }
 
     func receive(_ input: Input) -> Subscribers.Demand {
-        return _downstream.receive(_transform(input))
-    }
-
-    func receive(completion: Subscribers.Completion<Failure>) {
-        _downstream.receive(completion: completion)
+        switch _genericTransform(input) {
+        case .success(let output):
+            return downstream.receive(output)
+        case .failure(let error):
+            downstream.receive(completion: .failure(error))
+            return .none
+        }
     }
 
     func request(_ demand: Subscribers.Demand) {
-        _upstreamSubscription?.request(demand)
+        upstreamSubscription?.request(demand)
     }
 
-    func cancel() {
-        _upstreamSubscription?.cancel()
-        _upstreamSubscription = nil
+    override func cancel() {
+        upstreamSubscription?.cancel()
     }
 }
 
-// MARK: - CustomStringConvertible
-extension Inner: CustomStringConvertible {
-    public var description: String { return "Map" }
+extension Publishers.Map {
+
+    private final class Inner<Downstream: Subscriber>
+        : _Map<Upstream, Downstream>,
+          Subscriber
+        where Downstream.Failure == Upstream.Failure
+    {
+        init(downstream: Downstream,
+             transform: @escaping (Upstream.Output) -> Downstream.Input) {
+            super.init(downstream: downstream,
+                       genericTransform: { input in .success(transform(input)) })
+        }
+
+        func receive(completion: Subscribers.Completion<Failure>) {
+            downstream.receive(completion: completion)
+        }
+    }
 }
 
-// MARK: - CustomReflectable
-extension Inner: CustomReflectable {
-    public var customMirror: Mirror { return Mirror(self, children: EmptyCollection()) }
+extension Publishers.TryMap {
+
+    private final class Inner<Downstream: Subscriber>
+        : _Map<Upstream, Downstream>,
+          Subscriber
+        where Downstream.Failure == Error
+    {
+        init(downstream: Downstream,
+             transform: @escaping (Upstream.Output) throws -> Downstream.Input) {
+            super.init(downstream: downstream,
+                       genericTransform: { input in Result { try transform(input) } })
+        }
+
+        func receive(completion: Subscribers.Completion<Failure>) {
+            switch completion {
+            case .finished:
```

Comment:
It doesn't look like you're transforming the completion here.  Can this just be `downstream.receive(completion: completion)` instead of the switch?

Reply:
No, the types won't match: the `completion` argument has the type `Subscribers.Completion<Upstream.Failure>`, and the `downstream.receive(completion:)` method expects `Subscribers.Completion<Error>`.

Since user-defined generic types in Swift are invariant relative to their generic parameters,  `Subscribers.Completion<Upstream.Failure>` is not a subtype of `Subscribers.Completion<Error>` and therefore cannot be passed where the latter is expected

Reply:
I see, that's a bit annoying that we can see it should be that way but that the type system can't.